include/crocoddyl/multibody/actions/free-fwddyn.hxx

```diff
@@ -154,12 +154,14 @@ void DifferentialActionModelFreeFwdDynamicsTpl<Scalar>::quasiStatic(
   // Check the velocity input is zero
   assert_pretty(x.tail(state_->get_nv()).isZero(), "The velocity input should be zero for quasi-static to work.");
 
-  d->pinocchio.tau =
-      pinocchio::rnea(pinocchio_, d->pinocchio, q, VectorXs::Zero(state_->get_nv()), VectorXs::Zero(state_->get_nv()));
-
   d->tmp_xstatic.head(state_->get_nq()) = q;
-  actuation_->calc(d->multibody.actuation, d->tmp_xstatic, VectorXs::Zero(nu_));
-  actuation_->calcDiff(d->multibody.actuation, d->tmp_xstatic, VectorXs::Zero(nu_));
+  d->tmp_xstatic.tail(state_->get_nv()).setZero();
+  u.setZero();
+
+  pinocchio::rnea(pinocchio_, d->pinocchio, q, d->tmp_xstatic.tail(state_->get_nv()),
+                  d->tmp_xstatic.tail(state_->get_nv()));
+  actuation_->calc(d->multibody.actuation, d->tmp_xstatic, u);
+  actuation_->calcDiff(d->multibody.actuation, d->tmp_xstatic, u);
 
   u.noalias() = pseudoInverse(d->multibody.actuation->dtau_du) * d->pinocchio.tau;
```

Comment:
There is a huge memory allocation here!

Reply:
Hey @jcarpent! thank you so much for the feedback.

Do you refer to the `pseudoInverse` function?
In any case, could you let me know how should I remove the memory allocation that you pointed out?

Below, I have attached the code of the `pseudoInverse` function:
https://github.com/loco-3d/crocoddyl/blob/master/include/crocoddyl/core/utils/math.hpp#L20-L54

Reply:
The memory allocation occurs within `pseudoInverse` (the SVD struct from Eigen) but also when returning the value par copy. The best action would be to store within the data struct the SVD struct together with the Matrix storage to contain the temporaries and the pseudo-inverse matrix.

Reply:
If I were you, I would keep the current signature, while also providing a new one which does not perform any allocation.

Reply:
> The memory allocation occurs within pseudoInverse (the SVD struct from Eigen)

I thought that the `static` keyword in the SVD structure is avoiding to allocate memory, am I right? 

> but also when returning the value par copy.

I see this point! If I could create another signature (for `pseudoInverse`)  were we pass the output inversion. And of course, we can store this output in the action-data structure.

> The best action would be to store within the data struct the SVD struct together with the Matrix storage to contain the temporaries and the pseudo-inverse matrix.

I agree that this could be a workaround. But if I am right above, then I would prefer to avoid storing the SVD structure in the action-data structure.

Reply:
> I thought that the `static` keyword in the SVD structure is avoiding to allocate memory, am I right?

Not at all. You are mixing the different possible use of the `static` keyword.

Reply:
> Not at all. You are mixing the different possible use cases of the static keyword.

Thanks again for the feedback.

> If I were you, I would keep the current signature, while also providing a new one which does not perform any allocation.

This might be possible by defining two different data structures, and then parsing internally it. It seems a complex solution, but it could be implemented if everyone like it. Note that we cannot create two signatures for `quasiStatic` as it is a virtual function (which, for instance, is used in `ShootingProblem`).

Would be fine for you if I handle the issue with the `pseudoInverse` mem-allocation in another PR? (request due to time constraints).

If you agree, then I will create an issue to do not forget this task :)

Reply:
Quasi-static is normally not called during runtime, the use is mostly for warm-starting the solver, so the fix would not be that urgent.

Reply:
I am closing this conversation after I have created the issue: https://github.com/loco-3d/crocoddyl/issues/981.

Thanks guys for the feedback. Let me know if there is something left in this PR.